src/TypeChecker/Inference.hs

```diff
@@ -0,0 +1,521 @@
+  typeVar   <- newTypeVar
+  let env'  = extendEnv env ((TypeVar id), (Scheme [] typeVar))
+  (sub,typ) <- infer expr env'
+  return  (sub, FunctionType (apply sub typeVar) typ)
+
+inferLambdaWildcard :: Pattern a -> Expr a -> Environment -> Inference (Substitution, Type)
+inferLambdaWildcard (WildcardPattern pos _ ) expr env = do
+  typeVar   <- newTypeVar
+  (sub,typ) <- infer expr env
+  return(sub, FunctionType typeVar typ)
+
+inferLambdaConstructor :: Pattern a -> Expr a -> Environment -> Inference (Substitution, Type)
+inferLambdaConstructor p@(ConstructorPattern name _ pos _ ) expr env = do
+  (envPattern, typePattern) <- inferPattern p env
+  (subExpr, typeExpr) <- infer expr envPattern
+  return (subExpr, FunctionType typePattern typeExpr)
+
+
+inferLambdaTuple :: Pattern a -> Expr a -> Environment -> Inference (Substitution, Type)
+inferLambdaTuple pattern expr env = do
+  (envPattern, typePattern) <- inferPattern pattern env
+  (subExpr, typeExpr)       <- infer expr envPattern
+  return(subExpr, FunctionType (apply subExpr typePattern) typeExpr)
+
+inferLambdaNamed :: Pattern a -> Expr a -> Environment -> Inference (Substitution, Type)
+inferLambdaNamed (NamedPattern (Identifier id) pattern pos _ ) expr env = do
+  (envPattern, typePattern) <- inferPattern pattern env
+  (subExpr, typeExpr)       <- infer expr (extendEnv envPattern (TypeVar id, Scheme (Set.toList (freeTypeVar typePattern)) typePattern))
+  return(subExpr, FunctionType (apply subExpr typePattern) typeExpr)
+
+
+inferPrimitive :: PrimExpr -> Inference (Substitution, Type)
+inferPrimitive expr = case expr of
+  IntLit    _ -> return(emptySubstitution, PrimitiveType IntT)
+  CharLit   _ -> return(emptySubstitution, PrimitiveType CharT)
+  FloatLit  _ -> return(emptySubstitution, PrimitiveType FloatT)
+  StringLit _ -> return(emptySubstitution, PrimitiveType StringT)
+  BoolLit   _ -> return(emptySubstitution, PrimitiveType BoolT)
+  UnitLit     -> return(emptySubstitution, PrimitiveType UnitT)
+
+
+inferVar :: Identifier -> Environment -> Inference (Substitution, Type)
+inferVar (Identifier id) (Environment env) = 
+  case Map.lookup (TypeVar id) env of
+    Nothing -> throwError $ "Unbound variable: " ++ T.unpack (prettyPrint 0 (TypeVariable (TypeVar id)))
+    Just scheme  -> do 
+      typeOfVar <- instantiate scheme
+      return (emptySubstitution, typeOfVar) 
+
+
+inferConstructor :: ConstructorName -> Environment -> Inference (Substitution, Type)
+inferConstructor (ConstructorName name) (Environment env) = 
+  case Map.lookup (TypeVar name) env of
+    Nothing -> throwError $ "Undeclared type constructor: " ++ show name
+    Just scheme -> do
+      typeOfConstructor <- instantiate scheme
+      return (emptySubstitution, typeOfConstructor)
+
+inferExpr :: Environment -> Expr a -> Inference Type
+inferExpr env e =
+    do  (s, t) <- infer e env
+        return (apply s t)
+
+
+-- TODO: Observation: Either environment should be extended 
```

Comment:
To accommodate custom types all we need is to generate functions from type constructors. Then these functions can be placed in the environment like any other identifiers.

```
type Maybe a where
  case Some a
  case None
```
would spawn two variables
```
Some : a -> Maybe a
None : Maybe a
```

However it would be convenient to store information about defined types alongside the environment for two reasons:
- We need the number of constructors to decide how big should their tags be (numbers which discriminate union members) and to check if the pattern matching is exhaustive
- We will need kind typechecking in the future

I propose that instead of just one Map there should be two Maps, one for identifiers and one for types which would be passed together as a pair.